File: src/main/java/br/ufsc/bridge/platform/validation/form/errors/FormErrorImpl.java
package br.ufsc.bridge.platform.validation.form.errors;

import java.time.LocalDate;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.function.BiConsumer;
import java.util.stream.Collectors;

import br.ufsc.bridge.metafy.MetaField;
import br.ufsc.bridge.metafy.MetaList;
import br.ufsc.bridge.platform.validation.engine.Rule;
import br.ufsc.bridge.platform.validation.rules.Rules;
import br.ufsc.bridge.platform.validation.util.Reflections;

public class FormErrorImpl implements FormError {

	private static final long serialVersionUID = -6446758048046334551L;

	private transient Object target;
	private String rootError;
	private Map<String, Object> errors = new HashMap<>();

	public FormErrorImpl(Object target) {
		super();
		this.target = target;
	}

	@Override public void error(String mensagem) {
		this.rootError = mensagem;
	}

	private void fieldError(String campo, String mensagem) {
		if (mensagem != null) {
			this.errors.put(campo, mensagem);
		} else {
			this.errors.remove(campo);
		}
	}

	@Override
	public void fieldError(MetaField<?> field, String mensagem) {
		this.fieldError(field.getAlias(), mensagem);
	}

	protected Object getFieldValue(MetaField<?> field) {
		return Reflections.getValue(this.target, field.getAlias());
	}

	@Override
	public FormError formError(MetaField<?> field) {
		FormError error = (FormError) this.errors.get(field.getAlias());
		if (error == null) {
			error = new FormErrorImpl(this.getFieldValue(field));
			this.errors.put(field.getAlias(), error);
		}
		return error;
	}

	@Override
	public ListError listError(MetaList<?> field) {
		ListError error = (ListError) this.errors.get(field.getAlias());
		if (error == null) {
			error = new ListErrorImpl((List<?>) this.getFieldValue(field));
			this.errors.put(field.getAlias(), error);
		}
		return error;
	}

	@Override
	public boolean isValid(MetaField<?> field) {
		boolean valid = true;
		Object error = this.errors.get(field.getAlias());
		if (error instanceof String) {
			valid = false;
		} else if (error instanceof ValidationError) {
			valid = ((ValidationError) error).isValid();
		}
		return valid;
	}

	@Override
	public boolean isValid() {
		boolean valid = this.rootError == null;

		if (valid) {
			Iterator<Object> iterator = this.errors.values().iterator();
			while (iterator.hasNext()) {
				Object validationError = iterator.next();
				if (validationError instanceof ValidationError) {
					boolean childFormValid = ((ValidationError) validationError).isValid();
					valid = valid && childFormValid;
					if (childFormValid) {
						iterator.remove();
					}
				} else {
					valid = false;
				}
			}
		}

		return valid;
	}

	@Override public Object getErrors() {
		if (this.rootError != null) {
			return this.rootError;
		}

		return this.errors.entrySet()
				.stream()
				.collect(Collectors.toMap(entry -> entry.getKey(), entry -> {
					Object value = entry.getValue();
					if (value instanceof ValidationError) {
						return ((ValidationError) value).getErrors();
					}
					return value;
				}));
	}

	@Override
	public <F> FormError check(MetaField<F> field, Rule<F> rule) {
		if (this.isValid(field)) {
			String result;
			if (this.target != null) {
				result = rule.validate((F) Reflections.getValue(this.target, field.getAlias()));
			} else {
				result = rule.validate(null);
			}
			this.fieldError(field, result);
		}
		return this;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result + (this.target == null ? 0 : this.target.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (!super.equals(obj)) {
			return false;
		}
		if (this.getClass() != obj.getClass()) {
			return false;
		}
		FormErrorImpl other = (FormErrorImpl) obj;
		if (this.target == null) {
			if (other.target != null) {
				return false;
			}
		} else if (!this.target.equals(other.target)) {
			return false;
		}
		return true;
	}

	// Regras
	@Override
	public void alfaNumerico(MetaField<String> field) {
		this.check(field, Rules.alfaNumerico);
	}

	@Override
	public void beforeToday(MetaField<LocalDate> field) {
		this.check(field, Rules.beforeToday);
	}

	@Override
	public void cep(MetaField<String> field) {
		this.check(field, Rules.cep);
	}

	@Override
	public void cpf(MetaField<String> field) {
		this.check(field, Rules.cpf);
	}

	@Override
	public void cnpj(MetaField<String> field) {
		this.check(field, Rules.cnpj);
	}

	@Override
	public void cns(MetaField<String> field) {
		this.check(field, Rules.cns);
	}

	@Override
	public void email(MetaField<String> field) {
		this.check(field, Rules.email);
	}

	@Override
	public void empty(MetaField<?> field) {
		this.check(field, Rules.empty);
	}

	@Override
	public void hour(MetaField<String> field) {
		this.check(field, Rules.hour);
	}

	@Override
	public void length(MetaField<String> field, int length) {
		this.check(field, Rules.length(length));
	}

	@Override
	public void match(MetaField<String> field, String regex) {
		this.check(field, Rules.match(regex));
	}

	@Override
	public void maxLength(MetaField<String> field, int maxLength) {
		this.check(field, Rules.maxLength(maxLength));
	}

	@Override
	public <T extends Number> void maxRange(MetaField<T> field, T maxRange) {
		this.check(field, Rules.maxRange(maxRange));
	}

	@Override
	public void minLength(MetaField<String> field, int minLength) {
		this.check(field, Rules.minLength(minLength));
	}

	@Override
	public <T extends Number> void minRange(MetaField<T> field, T minRange) {
		this.check(field, Rules.minRange(minRange));
	}

	@Override
	public <T extends Number> void range(MetaField<T> field, T minRange, T maxRange) {
		this.check(field, Rules.range(minRange, maxRange));
	}

	@Override
	public void required(MetaField<?> field) {
		this.check(field, Rules.required);
	}

	@Override
	public void phone(MetaField<String> field) {
		this.check(field, Rules.phone);
	}

	@Override
	public <T> FormError forEach(MetaList<T> field, BiConsumer<T, FormError> itemValidator) {
		@SuppressWarnings("unchecked")
		List<T> list = (List<T>) this.getFieldValue(field);

		if (list == null || list.isEmpty()) {
			return this;
		}

		ListError listErrors = this.listError(field);

		for (int i = 0; i < list.size(); i++) {
			T item = list.get(i);
			itemValidator.accept(item, listErrors.itemError(i));
		}
		return this;
	}

}

File: src/main/kotlin/br/ufsc/bridge/platform/validation/form/errors/FormErrorImpl.kt
package br.ufsc.bridge.platform.validation.form.errors

import br.ufsc.bridge.metafy.MetaField
import br.ufsc.bridge.metafy.MetaList
import br.ufsc.bridge.platform.validation.engine.Rule
import br.ufsc.bridge.platform.validation.util.Reflections
import java.util.function.BiConsumer
import kotlin.reflect.KProperty
import kotlin.reflect.KProperty1

class FormErrorImpl<T> @JvmOverloads constructor(
        private val target: T? = null,
        private val internalErrors: HashMap<String, Any> = HashMap(),
        private var rootError: String? = null
) : FormError<T> {
    override val isValid: Boolean
        get() {
            var valid = this.rootError == null

            if (valid) {
                val iterator = this.internalErrors.values.iterator()
                while (iterator.hasNext()) {
                    val validationError = iterator.next()
                    valid = if (validationError is ValidationError) {
                        val childFormValid = validationError.isValid
                        if (childFormValid) {
                            iterator.remove()
                        }
                        valid && childFormValid
                    } else {
                        false
                    }
                }
            }

            return valid
        }

    override fun isValid(field: MetaField<*>): Boolean {
        var valid = true
        val error = this.internalErrors[field.alias]
        if (error is String) {
            valid = false
        } else if (error is ValidationError) {
            valid = error.isValid
        }
        return valid
    }

    override fun <R> isValid(property: KProperty1<T, R>): Boolean {
        var valid = true
        val error = this.internalErrors[property.name]
        if (error is String) {
            valid = false
        } else if (error is ValidationError) {
            valid = error.isValid
        }
        return valid
    }

    override fun fieldError(property: KProperty<*>, mensagem: String?) {
        if (mensagem != null) {
            this.internalErrors[property.name] = mensagem
        } else {
            this.internalErrors.remove(property.name)
        }
    }

    override fun fieldError(field: MetaField<*>, mensagem: String?) {
        if (mensagem != null) {
            this.internalErrors[field.alias] = mensagem
        } else {
            this.internalErrors.remove(field.alias)
        }
    }

    protected fun getFieldValue(field: MetaField<*>): Any? {
        return if (null != target) Reflections.getValue(this.target, field.alias) else null
    }

    protected fun <R> getPropertyValue(property: KProperty1<T, R>): R? {
        return if (null != target) property.call(target) else null
    }

    override fun error(mensagem: String?) {
        rootError = mensagem
    }

    override fun <E> formError(field: MetaField<E>): FormError<E> {
        return this.internalErrors.getOrPut(field.alias, { FormErrorImpl(getFieldValue(field)) }) as FormError<E>
    }

    override fun <E> formError(field: MetaField<E>, error: FormError<E>) {
        this.internalErrors.putIfAbsent(field.alias, error)
    }

    override fun <R> formError(property: KProperty1<T, R>): FormError<R> {
        return this.internalErrors.getOrPut(property.name, { FormErrorImpl(getPropertyValue(property)) }) as FormError<R>
    }

    override fun <R> formError(property: KProperty1<T, R?>, error: FormError<R>) {
        this.internalErrors.putIfAbsent(property.name, error)
    }

    override fun <E> listError(field: MetaList<E>): ListError<E> {
        return this.internalErrors.getOrPut(field.alias, { ListErrorImpl(getFieldValue(field) as List<E>?) }) as ListError<E>
    }

    override fun <R : List<E>?, E> listError(property: KProperty1<T, R>): ListError<E> {
        return this.internalErrors.getOrPut(property.name, { ListErrorImpl(getPropertyValue(property)) }) as ListError<E>
    }

    override fun <F : Any?> check(field: MetaField<F>, rule: Rule<F>): FormError<T> {
        if (isValid(field)) {
            val result = rule.validate(getFieldValue(field) as F?)
            fieldError(field, result)
        }
        return this
    }

    override fun <R> check(property: KProperty1<T, R>, rule: Rule<R>): FormError<T> {
        if (isValid(property)) {
            val result = rule.validate(getPropertyValue(property))
            fieldError(property, result)
        }
        return this
    }

    override fun <E : Any?> forEach(field: MetaList<E>, itemValidator: BiConsumer<E, FormError<E>>): FormError<T> {
        val list = getFieldValue(field) as List<*>?

        if (list == null || list.isEmpty()) {
            return this
        }

        val listErrors = this.listError(field)

        for (i in list.indices) {
            val item = list[i]
            itemValidator.accept(item as E, listErrors.itemError(i))
        }
        return this
    }

    override fun <R : List<E>?, E : Any?> forEach(property: KProperty1<T, R>, itemValidator: (E, FormError<E>) -> Unit): FormError<T> {
        val list = property.call(target)

        if (list == null || list.isEmpty()) {
            return this
        }

        val listErrors = this.listError(property)

        for (i in list.indices) {
            val item = list[i]
            itemValidator(item, listErrors.itemError(i))
        }
        return this
    }

    override val errors: Any
        get() {
            if (this.rootError != null) {
                return this.rootError as String
            } else {
                fun convertValue(value: Any): Any {
                    var convertedValue: Any = value
                    if (value is ValidationError) {
                        convertedValue = value.errors
                    }
                    return convertedValue
                }

                return this.internalErrors.entries.associateBy({ entry -> entry.key }, { entry -> convertValue(entry.value) })
            }
        }

}